### ISSUE TITLE
Test failures on CentOS 6.

### ISSUE THREAD

**Issue by dumol** [Legrandin/pycryptodome#187]
Using latest master, currently 61abaacbd29ec430a43d804e43f041bcb3c16c4c (but reproducible with 3.6.3 as well), lots of tests fail on CentOS 6.10 with latest updates installed (the same was true for previous version: 6.9).

```
Ran 39868 tests in 81.428s

FAILED (failures=911, errors=23481)
```

If running only the faster set with `setup.py test --skip-slow-tests`:

```
Ran 6383 tests in 34.978s

FAILED (failures=839, errors=3)
```

Some examples:

```
======================================================================
ERROR: test_bytearray (Crypto.SelfTest.Cipher.test_OCB.OcbTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "build/lib.linux-x86_64-2.7/Crypto/SelfTest/Cipher/test_OCB.py", line 267, in test_bytearray
    pt_test = cipher4.decrypt_and_verify(bytearray(ct_test), bytearray(tag_test))
  File "build/lib.linux-x86_64-2.7/Crypto/Cipher/_mode_ocb.py", line 488, in decrypt_and_verify
    self.verify(received_mac_tag)
  File "build/lib.linux-x86_64-2.7/Crypto/Cipher/_mode_ocb.py", line 440, in verify
    raise ValueError("MAC check failed")
ValueError: MAC check failed

======================================================================
ERROR: testImportKey8 (Crypto.SelfTest.PublicKey.test_import_DSA.ImportKeyTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "build/lib.linux-x86_64-2.7/Crypto/SelfTest/PublicKey/test_import_DSA.py", line 274, in testImportKey8
    key_obj = DSA.importKey(pem, "PWDTEST")
  File "build/lib.linux-x86_64-2.7/Crypto/PublicKey/DSA.py", line 646, in import_key
    (der, marker, enc_flag) = PEM.decode(tostr(extern_key), passphrase)
  File "build/lib.linux-x86_64-2.7/Crypto/IO/PEM.py", line 164, in decode
    data = unpad(objdec.decrypt(data), objdec.block_size)
  File "build/lib.linux-x86_64-2.7/Crypto/Util/Padding.py", line 93, in unpad
    raise ValueError("PKCS#7 padding is incorrect.")
ValueError: PKCS#7 padding is incorrect.

======================================================================
ERROR: test3 (Crypto.SelfTest.IO.test_PKCS8.PKCS8_Decrypt)
Verify unwrapping with encryption
----------------------------------------------------------------------
Traceback (most recent call last):
  File "build/lib.linux-x86_64-2.7/Crypto/SelfTest/IO/test_PKCS8.py", line 392, in test3
    res1, res2, res3 = PKCS8.unwrap(t[4], b("TestTest"))
  File "build/lib.linux-x86_64-2.7/Crypto/IO/PKCS8.py", line 186, in unwrap
    raise ValueError("Error decoding PKCS#8 (%s)" % error_str)
ValueError: Error decoding PKCS#8 (PBES1[Unknown OID for PBES1],PBES2[Invalid])

======================================================================
FAIL: runTest (Crypto.SelfTest.Cipher.common.ByteArrayTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "build/lib.linux-x86_64-2.7/Crypto/SelfTest/Cipher/common.py", line 299, in runTest
    self.assertEqual(self.ciphertext, ct)  # encrypt
AssertionError: '69c4e0d86a7b0430d8cdb78070b4c55a' != '13cfe27f30bd686f561940dad0f067e2'

======================================================================
FAIL: runTest (Crypto.SelfTest.Cipher.common.CipherSelfTest)
AESNI #1: FIPS 197 C.1 (AES-128)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "build/lib.linux-x86_64-2.7/Crypto/SelfTest/Cipher/common.py", line 137, in runTest
    self.assertEqual(self.ciphertext, ct)  # encrypt
AssertionError: '69c4e0d86a7b0430d8cdb78070b4c55a' != '13cfe27f30bd686f561940dad0f067e2'

======================================================================
FAIL: runTest (Crypto.SelfTest.Cipher.common.CipherSelfTest)
```

Just in case, stderr output from building with included gcc 4.4.7:
```
Testing support for 128-bit integer
Target does support 128-bit integer
Testing support for intrin.h header
Target does not support intrin.h header
Testing support for cpuid.h header
Target does support cpuid.h header
Testing support for posix_memalign
Target does support posix_memalign
Testing support for AESNI intrinsics
Target does support AESNI intrinsics
Compiling support for AESNI instructions
Testing support for CLMUL intrinsics
Target does support CLMUL intrinsics
Compiling support for CLMUL instructions
```
```
src/AESNI.c: In function 'sub_rot':
src/AESNI.c:90: warning: 'result' may be used uninitialized in this function
```
```
rc/scrypt.c: In function 'scryptBlockMix':
src/scrypt.c:91: warning: assignment from incompatible pointer type
src/scrypt.c: In function 'scryptROMix':
src/scrypt.c:136: warning: passing argument 1 of 'scryptBlockMix' from incompatible pointer type
src/scrypt.c:76: note: expected 'const uint8_t (*)[64]' but argument is of type 'uint8_t (*)[64]'
src/scrypt.c:150: warning: passing argument 1 of 'scryptBlockMix' from incompatible pointer type
src/scrypt.c:76: note: expected 'const uint8_t (*)[64]' but argument is of type 'uint8_t (*)[64]'
```

RHEL 6.x is still supported for a number of years, until June 30 2024 according to https://en.wikipedia.org/wiki/Red_Hat_Enterprise_Linux#Product_life_cycle.

**Comment by Legrandin** [comment]
Does it work if you install the Linux wheel of 3.6.3 with pip and then run `python27 -m Crypto.SelfTest`?

I can't reproduce it with CentOS 6.10 64-bit, with Python 2.7.13 and gcc 4.4.7.

The strangest thing is that I don't get that warning on AESNI.c, which might indicate that - at your side - the SSE2 intrinsics are compiled out (and with those, AES test cases and a bunch of others would fail).

If you don't mind, can you also paste how the modules get compiled by gcc?

**Comment by dumol** [comment]
Using latest upstream wheel (now 3.6.4) works just fine, thank you! For the reference: https://files.pythonhosted.org/packages/1d/4f/60b5bb34888dffdb0fc8db6740b5a037bf039570a37b8f5173ee24a33f32/pycryptodome-3.6.4-cp27-cp27m-manylinux1_x86_64.whl

But maybe we can get to the bottom of this? We can't use wheels everywhere because we have to support a few exotic platforms. Even on Linux, we also support Alpine, which uses Musl instead of Glibc. By the way, I skimmed the docs, but couldn't find an explanation, what is the difference between `pycryptodome-3.6.4-cp27-cp27m-manylinux1_x86_64.whl` and `pycryptodome-3.6.4-cp27-cp27mu-manylinux1_x86_64.whl`? 

Back to build issues, this is how `AESNI.c` module is compiled on our side, as an example:

```
building 'Crypto.Cipher._raw_aesni' extension
gcc -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -fPIC -DHAVE_X86INTRIN_H -DPYCRYPTO_LITTLE_ENDIAN -DLTC_NO_ASM -DHAVE_UINT128 -DHAVE_CPUID_H -DHAVE_POSIX_MEMALIGN -Isrc/ -I/srv/buildslave/runtime/build-rhel6-x64/slave/python-package-rhel-6/build/build/python2.7-rhel6-x64/include/python2.7 -c src/AESNI.c -o build/temp.linux-x86_64-2.7/src/AESNI.o -O3 -msse2 -maes
src/AESNI.c: In function 'sub_rot':
src/AESNI.c:90: warning: 'result' may be used uninitialized in this function
```

Maybe you can try once more our Python package at https://github.com/chevah/python-package. Just checkout the `4985-pycryptodome-instead-of-pycrypto` branch and issue a `./chevah build`. 

I hope it will reproduce, we have two CentOS 6.10 VMs which exhibit this issue. If not, it would still be interesting to know. Although there's nothing special with our VMs, maybe there's something wrong with our installations / build process / etc. Thank you once more!

**Comment by dumol** [comment]
Beware that I'll switch to using your upstream wheel on RHEL 6 in our `python-package` repo. Please use revision https://github.com/chevah/python-package/commit/025eff6b6912e1cd7cda3ebe00e0f0c5b5a839d0 if you try reproducing this as detailed in the comment above. Thanks!

Just a side note: this also works around unrelated build issues we currently have with SLES 11 SP4 (with or without the Security Module). To be reported separately if there is interest…

**Comment by Legrandin** [comment]
Thanks for providing these details: it will allow me to dig deeper, but you will hear again from me only in a few weeks...

For this:

> what is the difference between pycryptodome-3.6.4-cp27-cp27m-manylinux1_x86_64.whl and pycryptodome-3.6.4-cp27-cp27mu-manylinux1_x86_64.whl?

The difference relates to which flags were used to compile CPython in Windows: the presence of the latter `u` after `cp27` means that `--enable-unicode=ucs4` (wide-unicode) was passed, the absence indicates `--enable-unicode=ucs4` (narrow-unicode) instead.

**Comment by dumol** [comment]
I'll probably merge in master and delete the above referenced branch, but I'll take care to create one for this issue, to easily reproduce the problem. Will come back here with details, if/when that happens.

Thank you for the wheels explanation. And, if my assumptions are right, _have a nice vacation!_

**Comment by Legrandin** [comment]
Fixed in v3.6.5

**Comment by dumol** [comment]
I can confirm v3.6.5 fixes this issue for us too. Thank you once more… Great job! ;-]
